sympy/core/kind.py

```diff
@@ -385,3 +386,94 @@ def __doc__(self):
             docs.append(s)
 
         return '\n\n'.join(docs)
+
+
+class PropertyDispatcher:
```

Comment:
Do you think it's wise to introduce new ad-hoc dispatchers? I'm pointing this out because it could make analyzing the code a bit harder.

We have already multipledispatching, can we use that in order to dispatch on both `kind` and `expr`?

As an alternative, can't we have a single-dispatcher over the kinds, returning a shape getter?
```python
def shape(expr):
    get_shape = shape_of_kind(expr.kind)
    return get_shape(expr)

@shape_of_kind.register
def _(kind: NumberKind):
    return lambda expr: ()

@shape_of_kind.register
def _(kind: ArrayKind):
    return get_shape_for_array

@get_shape_for_array.register
def _(expr: Integral)
    return shape(expr.args[0])
```

Another possibility would be using the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern)
```python
def shape(expr):
    return accept_shape(expr.kind, expr)

@accept_shape.register
def _(kind: NumberKind, expr):
    return visit_shape_numberkind(expr)

@accept_shape.register
def _(kind: ArrayKind, expr)
    return visit_shape_arraykind(expr)

@visit_shape_numberkind.register
def _(expr: Integral):
    return ()  # or None or NotImplementedError() if you prefer

@visit_shape_arraykind.register
def _(expr: Integral):
    return shape(expr.args[0])
```

The second approach has the advantage of being translatable into C++, which is theory could be used to port the code to [SymEngine](https://github.com/symengine/symengine) if we're going to replace parts of SymPy with C++ code in the future.

Reply:
Thanks. I was uneasy with `PropertyDispatcher`, and this looks better.